pkg/codec/property_extractor_test.go

```diff
@@ -0,0 +1,250 @@
+)
+
+func TestPropertyExtractor(t *testing.T) {
+	t.Parallel()
+
+	type testStruct struct {
+		A bool
+		B int64
+	}
+
+	type nestedTestStruct struct {
+		A string
+		B testStruct
+	}
+
+	onChainType := reflect.TypeOf(nestedTestStruct{})
+
+	extractor := codec.NewPropertyExtractor("A")
+	invalidExtractor := codec.NewPropertyExtractor("A.B")
+	nestedExtractor := codec.NewPropertyExtractor("B.B")
+
+	t.Run("RetypeToOffChain sets the type for offchain to the onchain property", func(t *testing.T) {
+		offChainType, err := extractor.RetypeToOffChain(reflect.TypeOf(nestedTestStruct{}), "")
+		require.NoError(t, err)
+		require.Equal(t, reflect.TypeOf(""), offChainType)
+	})
+
+	t.Run("RetypeToOffChain works on pointers", func(t *testing.T) {
+		offChainType, err := extractor.RetypeToOffChain(reflect.TypeOf(&nestedTestStruct{}), "")
+		require.NoError(t, err)
+
+		str := ""
+
+		require.Equal(t, reflect.TypeOf(&str), offChainType)
+	})
+
+	t.Run("RetypeToOffChain works on slices", func(t *testing.T) {
+		offChainType, err := extractor.RetypeToOffChain(reflect.TypeOf([]nestedTestStruct{}), "")
+		require.NoError(t, err)
+
+		require.Equal(t, reflect.TypeOf([]string{}), offChainType)
+	})
+
+	t.Run("RetypeToOffChain works on arrays", func(t *testing.T) {
+		arrayLen := 3
+		offChainType, err := extractor.RetypeToOffChain(reflect.ArrayOf(arrayLen, onChainType), "")
+		require.NoError(t, err)
+
+		require.Equal(t, reflect.Array, offChainType.Kind())
+		require.Equal(t, arrayLen, offChainType.Len())
+	})
+
+	t.Run("RetypeToOffChain returns error for missing field", func(t *testing.T) {
+		_, err := invalidExtractor.RetypeToOffChain(reflect.TypeOf(nestedTestStruct{}), "")
+		require.NotNil(t, err)
+	})
+
+	t.Run("TransformToOnChain and TransformToOffChain works on structs", func(t *testing.T) {
+		_, err := extractor.RetypeToOffChain(reflect.TypeOf(nestedTestStruct{}), "")
+		require.NoError(t, err)
+
+		onChainValue := nestedTestStruct{
+			A: "test",
+			B: testStruct{
+				A: true,
+				B: 42,
+			},
+		}
+
+		offChainValue, err := extractor.TransformToOffChain(onChainValue, "")
+		require.NoError(t, err)
+		require.Equal(t, "test", offChainValue)
+
+		lossyOnChain, err := extractor.TransformToOnChain("value", "")
+		require.NoError(t, err)
+
+		expected := nestedTestStruct{
+			A: "value",
+			B: testStruct{
+				A: false,
+				B: 0,
+			},
+		}
+
+		assert.Equal(t, expected, lossyOnChain)
+	})
+
+	t.Run("TransformToOnChain and TransformToOffChain works on nested structs", func(t *testing.T) {
+		_, err := nestedExtractor.RetypeToOffChain(reflect.TypeOf(nestedTestStruct{}), "")
+		require.NoError(t, err)
+
+		onChainValue := nestedTestStruct{
+			A: "test",
+			B: testStruct{
+				A: true,
+				B: 42,
+			},
+		}
+
+		offChainValue, err := nestedExtractor.TransformToOffChain(onChainValue, "")
+		require.NoError(t, err)
+		require.Equal(t, int64(42), offChainValue)
+
+		lossyOnChain, err := nestedExtractor.TransformToOnChain(int64(3), "")
```

Comment:
nit: it's nice to use the same value that you got from the result to show it works with the other transform.

Reply:
Like using the `offChainValue` to transform to `lossyOnChain`?

Reply:
yeah